File: python/find-sessions.py
#!/usr/bin/env python3
# find-sessions.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway1:skills/tmux/scripts/find-sessions.sh
# auch in: OpenClaw@gateway2:skills/tmux/scripts/find-sessions.sh
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

import os
import sys
import subprocess
import argparse
from pathlib import Path

def usage():
    print("""
Usage: find-sessions.sh [-L socket-name|-S socket-path|-A] [-q pattern]

List tmux sessions on a socket (default tmux socket if none provided).

Options:
  -L, --socket       tmux socket name (passed to tmux -L)
  -S, --socket-path  tmux socket path (passed to tmux -S)
  -A, --all          scan all sockets under CLAWDBOT_TMUX_SOCKET_DIR
  -q, --query        case-insensitive substring to filter session names
  -h, --help         show this help
""")

def list_sessions(label, tmux_args, query):
    tmux_cmd = ["tmux"] + tmux_args + ["list-sessions", "-F", "#{session_name}\t#{session_attached}\t#{session_created_string}"]
    
    try:
        result = subprocess.run(tmux_cmd, capture_output=True, text=True)
        if result.returncode != 0:
            print(f"No tmux server found on {label}", file=sys.stderr)
            return False
    except FileNotFoundError:
        print("tmux not found in PATH", file=sys.stderr)
        sys.exit(1)
    
    sessions = result.stdout.strip()
    
    if query:
        lines = sessions.split('\n') if sessions else []
        filtered_lines = []
        for line in lines:
            if query.lower() in line.lower():
                filtered_lines.append(line)
        sessions = '\n'.join(filtered_lines)
    
    if not sessions:
        print(f"No sessions found on {label}")
        return True
    
    print(f"Sessions on {label}:")
    for line in sessions.split('\n'):
        if line.strip():
            parts = line.split('\t')
            if len(parts) >= 3:
                name, attached, created = parts[0], parts[1], parts[2]
                attached_label = "attached" if attached == "1" else "detached"
                print(f"  - {name} ({attached_label}, started {created})")
    return True

def main():
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('-L', '--socket', dest='socket_name')
    parser.add_argument('-S', '--socket-path', dest='socket_path')
    parser.add_argument('-A', '--all', action='store_true')
    parser.add_argument('-q', '--query')
    parser.add_argument('-h', '--help', action='store_true')
    
    args, unknown = parser.parse_known_args()
    
    if args.help:
        usage()
        sys.exit(0)
    
    if unknown:
        print(f"Unknown option: {unknown[0]}", file=sys.stderr)
        usage()
        sys.exit(1)
    
    socket_name = args.socket_name or ""
    socket_path = args.socket_path or ""
    query = args.query or ""
    scan_all = args.all
    
    if scan_all and (socket_name or socket_path):
        print("Cannot combine --all with -L or -S", file=sys.stderr)
        sys.exit(1)
    
    if socket_name and socket_path:
        print("Use either -L or -S, not both", file=sys.stderr)
        sys.exit(1)
    
    socket_dir = os.environ.get('CLAWDBOT_TMUX_SOCKET_DIR', 
                               os.path.join(os.environ.get('TMPDIR', '/tmp'), 'clawdbot-tmux-sockets'))
    
    if scan_all:
        socket_path_obj = Path(socket_dir)
        if not socket_path_obj.is_dir():
            print(f"Socket directory not found: {socket_dir}", file=sys.stderr)
            sys.exit(1)
        
        sockets = list(socket_path_obj.iterdir())
        
        if not sockets:
            print(f"No sockets found under {socket_dir}", file=sys.stderr)
            sys.exit(1)
        
        exit_code = 0
        for sock in sockets:
            if sock.is_socket():
                if not list_sessions(f"socket path '{sock}'", ['-S', str(sock)], query):
                    exit_code = 1
        sys.exit(exit_code)
    
    tmux_args = []
    socket_label = "default socket"
    
    if socket_name:
        tmux_args.extend(['-L', socket_name])
        socket_label = f"socket name '{socket_name}'"
    elif socket_path:
        tmux_args.extend(['-S', socket_path])
        socket_label = f"socket path '{socket_path}'"
    
    list_sessions(socket_label, tmux_args, query)

if __name__ == "__main__":
    main()

File: javascript/find-sessions.js
#!/usr/bin/env node
// find-sessions.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway1:skills/tmux/scripts/find-sessions.sh
// auch in: OpenClaw@gateway2:skills/tmux/scripts/find-sessions.sh
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

function usage() {
  process.stderr.write(`\
Usage: find-sessions.sh [-L socket-name|-S socket-path|-A] [-q pattern]

List tmux sessions on a socket (default tmux socket if none provided).

Options:
  -L, --socket       tmux socket name (passed to tmux -L)
  -S, --socket-path  tmux socket path (passed to tmux -S)
  -A, --all          scan all sockets under CLAWDBOT_TMUX_SOCKET_DIR
  -q, --query        case-insensitive substring to filter session names
  -h, --help         show this help
`);
}

let socketName = "";
let socketPath = "";
let query = "";
let scanAll = false;
const socketDir = process.env.CLAWDBOT_TMUX_SOCKET_DIR || 
                  (process.env.TMPDIR ? path.join(process.env.TMPDIR, 'clawdbot-tmux-sockets') : '/tmp/clawdbot-tmux-sockets');

const args = process.argv.slice(2);
let i = 0;

while (i < args.length) {
  const arg = args[i];
  
  switch (arg) {
    case '-L':
    case '--socket':
      if (i + 1 >= args.length) {
        process.stderr.write("Option requires an argument: " + arg + "\n");
        usage();
        process.exit(1);
      }
      socketName = args[i + 1];
      i += 2;
      break;
      
    case '-S':
    case '--socket-path':
      if (i + 1 >= args.length) {
        process.stderr.write("Option requires an argument: " + arg + "\n");
        usage();
        process.exit(1);
      }
      socketPath = args[i + 1];
      i += 2;
      break;
      
    case '-A':
    case '--all':
      scanAll = true;
      i++;
      break;
      
    case '-q':
    case '--query':
      if (i + 1 >= args.length) {
        process.stderr.write("Option requires an argument: " + arg + "\n");
        usage();
        process.exit(1);
      }
      query = args[i + 1];
      i += 2;
      break;
      
    case '-h':
    case '--help':
      usage();
      process.exit(0);
      break;
      
    default:
      process.stderr.write("Unknown option: " + arg + "\n");
      usage();
      process.exit(1);
  }
}

if (scanAll && (socketName || socketPath)) {
  process.stderr.write("Cannot combine --all with -L or -S\n");
  process.exit(1);
}

if (socketName && socketPath) {
  process.stderr.write("Use either -L or -S, not both\n");
  process.exit(1);
}

// Check if tmux exists
const tmuxCheck = spawnSync('which', ['tmux'], { encoding: 'utf8' });
if (tmuxCheck.status !== 0) {
  process.stderr.write("tmux not found in PATH\n");
  process.exit(1);
}

function listSessions(label, ...tmuxArgs) {
  const tmuxCmd = ['tmux', ...tmuxArgs, 'list-sessions', '-F', '#{session_name}\t#{session_attached}\t#{session_created_string}'];
  
  const result = spawnSync(tmuxCmd[0], tmuxCmd.slice(1), { encoding: 'utf8' });
  
  if (result.status !== 0) {
    process.stderr.write(`No tmux server found on ${label}\n`);
    return false;
  }
  
  let sessions = result.stdout.trim();
  
  if (query) {
    const lines = sessions.split('\n').filter(line => line.toLowerCase().includes(query.toLowerCase()));
    sessions = lines.join('\n');
  }
  
  if (!sessions) {
    console.log(`No sessions found on ${label}`);
    return true;
  }
  
  console.log(`Sessions on ${label}:`);
  
  sessions.split('\n').forEach(line => {
    if (!line) return;
    const [name, attached, created] = line.split('\t');
    const attachedLabel = attached === "1" ? "attached" : "detached";
    console.log(`  - ${name} (${attachedLabel}, started ${created})`);
  });
  
  return true;
}

if (scanAll) {
  if (!fs.existsSync(socketDir) || !fs.statSync(socketDir).isDirectory()) {
    process.stderr.write(`Socket directory not found: ${socketDir}\n`);
    process.exit(1);
  }
  
  const sockets = fs.readdirSync(socketDir)
    .map(file => path.join(socketDir, file))
    .filter(file => fs.statSync(file).isSocket());
  
  if (sockets.length === 0) {
    process.stderr.write(`No sockets found under ${socketDir}\n`);
    process.exit(1);
  }
  
  let exitCode = 0;
  for (const sock of sockets) {
    if (!listSessions(`socket path '${sock}'`, '-S', sock)) {
      exitCode = 1;
    }
  }
  process.exit(exitCode);
}

const socketLabel = socketName ? `socket name '${socketName}'` : 
                   socketPath ? `socket path '${socketPath}'` : 
                   "default socket";

const tmuxArgs = [];
if (socketName) {
  tmuxArgs.push('-L', socketName);
} else if (socketPath) {
  tmuxArgs.push('-S', socketPath);
}

listSessions(socketLabel, ...tmuxArgs);
